packages/halo-nonce-pcd/src/HaLoNoncePCD.ts

```diff
@@ -0,0 +1,196 @@
+   * The digest of the nonce + random string
+   */
+  rnd: StringArgument;
+
+  /**
+   * The signature of the nonce + random string
+   */
+  rndsig: StringArgument;
+}
+
+export interface HaLoNoncePCDClaim {
+  /**
+   * Incrementing nonce that was signed.
+   */
+  nonce: number;
+
+  /**
+   * Uncompressed hex string of the public key, starting with `04`
+   */
+  pubkeyHex: string;
+}
+
+export interface HaLoNoncePCDProof {
+  /**
+   * The signed digest, consisting of the nonce concatenated with a random string
+   */
+  signedDigest: string;
+
+  /**
+   * The cleaned signature of the nonce + random string,
+   */
+  cleanedSignature: string;
+}
+
+export class HaLoNoncePCD implements PCD<HaLoNoncePCDClaim, HaLoNoncePCDProof> {
+  type = HaLoNoncePCDTypeName;
+  claim: HaLoNoncePCDClaim;
+  proof: HaLoNoncePCDProof;
+  id: string;
+
+  public constructor(
+    id: string,
+    claim: HaLoNoncePCDClaim,
+    proof: HaLoNoncePCDProof
+  ) {
+    this.id = id;
+    this.claim = claim;
+    this.proof = proof;
+  }
+}
+
+export async function prove(args: HaLoNoncePCDArgs): Promise<HaLoNoncePCD> {
+  if (!args.requestNewPCD.value) {
+    if (
+      args.pk2.value === undefined ||
+      args.rnd.value === undefined ||
+      args.rndsig.value === undefined
+    ) {
+      throw new Error("Cannot make HaLoNoncePCD: some arguments are not set.");
+    }
+
+    const claim: HaLoNoncePCDClaim = {
+      nonce: parseInt(args.rnd.value.substring(0, 8)),
+      pubkeyHex: args.pk2.value,
+    };
+
+    if (isNaN(claim.nonce)) {
+      throw new Error("Nonce is not a valid number.");
+    }
+
+    try {
+      secp256k1.keyFromPublic(claim.pubkeyHex, "hex");
+    } catch (e) {
+      throw new Error("Unable to decode public key.");
+    }
+
+    // Clean up signature for easier verification
+    const sigBuf = Buffer.from(args.rndsig.value, "hex");
+    if (sigBuf.length < 2 || 2 + sigBuf[1] > sigBuf.length) {
+      throw new Error("Malformed signature in the rndsig field.");
+    }
+    const cutSig = sigBuf.subarray(0, 2 + sigBuf[1]).toString("hex");
+
+    const proof: HaLoNoncePCDProof = {
+      signedDigest: args.rnd.value,
+      cleanedSignature: cutSig,
+    };
+
+    return new HaLoNoncePCD(uuid(), claim, proof);
+  } else {
+    // TODO: directly use `execHaloCmdWeb` from the libhalo library, as in
+    // https://github.com/arx-research/libhalo/blob/master/docs/web-reactjs.md
+
+    return new HaLoNoncePCD(
+      uuid(),
+      { nonce: 0, pubkeyHex: "" },
+      { signedDigest: "", cleanedSignature: "" }
+    );
+  }
+}
+
+export async function verify(pcd: HaLoNoncePCD): Promise<boolean> {
+  // verify nonce in claim matches the one in the proof
+  if (pcd.claim.nonce !== parseInt(pcd.proof.signedDigest.substring(0, 8))) {
+    return false;
+  }
+
+  const rndBuf = Buffer.from(pcd.proof.signedDigest, "hex");
+  const hash = sha256.create();
+  const rndHashed = hash
+    .update(
+      Buffer.concat([
+        Buffer.from([0x19]),
```

Comment:
add a comment above this with a link to the spec

this looks EIP712 ish?

where does "Attest counter pk2" come from?

Reply:
It's just the prefix the HaLo guys added, good point on adding a link